gyronimo/dynamics/classical_boris.cc

```diff
@@ -31,169 +31,105 @@
-    const metric_covariant* Bmetric = Bfield->metric();
-    if (Emetric != Bmetric)
-      error(
-          __func__, __FILE__, __LINE__,
-          " Efield and Bfield must be in the same coordinate representation.",
-          1);
-  }
-
-  // ensure that the field metrics derive from metric_connected
+  if (!B) error(__func__, __FILE__, __LINE__, "no magnetic field.", 1);
+  if (E && B->metric() != E->metric())
+    error(__func__, __FILE__, __LINE__, "mismatched E/B coordinates.", 1);
   if (!metric_)
-    error(
-        __func__, __FILE__, __LINE__,
-        " field metrics must derive from 'gyronimo::metric_connected'.", 1);
-
-  // test if normalization factors are consistent
-  if (Efield) {
-    double E_m_factor = Efield->m_factor();
-    double B_m_factor = Bfield->m_factor();
-    if (std::abs(1.0 - E_m_factor / (Vref_ * B_m_factor)) > 1.0e-14)
-      error(
-          __func__, __FILE__, __LINE__,
-          " inconsistent Efield and Bfield normalization.", 1);
-  }
+    error(__func__, __FILE__, __LINE__, "field has no metric_connected.", 1);
 }
 
-//! Performs a time step `dt` update to the state `in` and returns the result.
+//! Returns the update of a state by a single time step `dt`.
 classical_boris::state classical_boris::do_step(
-    const state& in, const double& time, const double& dt) const {
-  // extract position and velocity from state
-  IR3 q_old = {in[0], in[1], in[2]};
-  IR3 v_old = {in[3], in[4], in[5]};
-  dIR3 e_old = my_morphism_->del(q_old);
-
-  // calculate fields
-  IR3 Efield = {0.0, 0.0, 0.0};
-  if (electric_field_) {
-    double Etime = time * iEfield_time_factor_;
-    IR3 E_contravariant = electric_field_->contravariant(q_old, Etime);
-    Efield = contraction<second>(e_old, E_contravariant);
-  }
-  double Btime = time * iBfield_time_factor_;
-  IR3 Bfield_contravariant = magnetic_field_->contravariant(q_old, Btime);
-  IR3 Bfield = contraction<second>(e_old, Bfield_contravariant);
-  double Bmag = std::sqrt(inner_product(Bfield, Bfield));
-  IR3 Bversor = Bfield / Bmag;
-
-  // perform cartesian_boris step
-  IR3 v_new = electric_field_
-      ? boris_push(v_old, Oref_, Efield, Bmag, Bversor, dt)
-      : boris_push(v_old, Oref_, Bmag, Bversor, dt);
-  // IR3 x_new = x_old + (Lref_*dt) * v_new;
-
-  // IR3 q_new = my_morphism_->inverse(x_new);
-  // IR3 u_new = my_morphism_->to_contravariant(v_new, q_new);
-
-  IR3 q_new = my_morphism_->translation(q_old, (Lref_ * dt) * v_new);
-
-  return {q_new[IR3::u], q_new[IR3::v], q_new[IR3::w],
-          v_new[IR3::u], v_new[IR3::v], v_new[IR3::w]};
+    const state& s, const double& time, const double& dt) const {
+  double Btime = time * iB_time_factor_;
+  const double tildeEref = Oref_ * iB_time_factor_ / (iE_time_factor_ * Vref_);
+
+  IR3 q = this->get_position(s), v = this->get_velocity(s);
+  double B = magnetic_field_->magnitude(q, Btime);
+  IR3 b = my_morphism_->from_contravariant(
+      magnetic_field_->contravariant_versor(q, Btime), q);
```

Comment:
Even though this is correct, I should point out why I wrote the code the other way.
Specifically when using a `helena` magnetic field, I noticed that the energy conservation wasn't being met to the accuracy (`~10^-6`) that I expected (`~10^-13`) from the Boris push - because the energy conservation during the rotation is analytical. I found that the problem was a mismatch between the `magnitude` and `covariant` interpolations of the `helena` equilibrium. So I computed the magnitude from the magnetic field vector itself to circumvent the problem.

Reply:
The problem you described is a symptom that the morphism functions and their derivatives (via bicubic interpolators in this case) are not fully compatible (up to machine precision) with the metric components. This is indeed a serious problem, but I believe it needs to be solved in the `{metric,morphism}_helena` family of objects rather than in the Boris steppers.